### ISSUE TITLE
Make NS1004 distinguish between non-virtual members and extension methods

### ISSUE THREAD

**Issue by petterh** [nsubstitute/NSubstitute.Analyzers#206]
Assume an interface and an extension method for that interface:

```csharp
public interface IHolder
{
    string Value { get; set; }
}

public static class IHolderExtensions
{
    public static void SetValue(this IHolder holder, string value) => holder.Value = value;
}
```

The following triggers an NS1004:

```csharp
string? value = null;
var holder = Substitute.For<IHolder>();
holder.SetValue(Arg.Do<string>(v => value = v)); // <==
holder.SetValue("hello");
value.Should().Be("hello");
```

With the provided definition of `SetValue`, I can ignore that; things work as expected because the call to the extension method basically boils down to this:

```csharp
holder.Value = Arg.Do<string>(v => value = v);
```

If I change `SetValue` to the following, things blow up, of course, so the analyzer is right to flag the issue:

```csharp
public static void SetValue(this IHolder holder, string value) => throw new InvalidOperationException("Go away");
```

## Proposal

If possible, I'd like to see the message

> Argument matcher used with a non-virtual member of a class.

changed to

> Argument matcher used with an extension method.

Might require a new rule.

**Comment by tpodolak** [comment]
Hi @petterh
We are planing to have separate rule for substitution of [extension methods](https://github.com/nsubstitute/NSubstitute.Analyzers/issues/170), so having separate rule for argument matchers makes sense. Thanks for pointing that out
